<a href="https://colab.research.google.com/github/Juliethh777/Inteligencia-Artificial-2./blob/main/Modelo_Predictivo_Casas_IA2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Modelo Predictivo de Precios de Casas

## Inteligencia Artificial 2

**Modelos utilizados:**
- Regresión Lineal
- Árbol de Decisión

**Dataset:** Housing Prices Dataset

## Objetivo

Desarrollar y comparar dos modelos predictivos para estimar el precio de las casas a partir de sus características.

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Importo las librerías porque necesito herramientas para organizar los datos,
# separar los datos de entrenamiento y prueba, crear los dos modelos
# y finalmente medir qué tan buenas son sus predicciones.

In [13]:
datos = pd.read_csv('/content/Housing.csv')

# Cargo el archivo Housing.csv y lo guardo en la variable "datos"
# para poder trabajar con la información de las casas.

In [3]:
datos.head()

# Muestro las primeras filas del archivo para comprobar que los datos
# se cargaron correctamente y conocer cómo están organizadas las columnas.

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished


In [4]:
columnas_si_no = [
    'mainroad',
    'guestroom',
    'basement',
    'hotwaterheating',
    'airconditioning',
    'prefarea'
]

for columna in columnas_si_no:
    datos[columna] = datos[columna].map({'yes': 1, 'no': 0})

# Convierto las respuestas "yes" y "no" en 1 y 0 porque
# los modelos necesitan trabajar con datos numéricos.

In [5]:
datos = pd.get_dummies(
    datos,
    columns=['furnishingstatus'],
    drop_first=True
)

# Convierto el tipo de amueblado en columnas numéricas,
# porque esta información también debe estar en un formato
# que los modelos puedan procesar.

In [6]:
X = datos.drop('price', axis=1)
y = datos['price']

# En X guardo las características de las casas que voy a utilizar
# para hacer las predicciones.
# En y guardo el precio, porque ese es el valor que quiero predecir.

In [7]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# Separo los datos en dos grupos:
# el 80% lo usaré para entrenar los modelos
# y el 20% restante para comprobar sus predicciones.

In [8]:
modelo_lineal = LinearRegression()

# Creo el modelo de regresión lineal porque quiero encontrar
# una relación entre las características de las casas y su precio.

modelo_lineal.fit(X_train, y_train)

# Entreno el modelo utilizando las casas que separé para aprender.

predicciones_lineal = modelo_lineal.predict(X_test)

# Utilizo el modelo entrenado para estimar el precio
# de las casas que dejé para hacer la prueba.

In [9]:
mae_lineal = mean_absolute_error(y_test, predicciones_lineal)
rmse_lineal = np.sqrt(mean_squared_error(y_test, predicciones_lineal))
r2_lineal = r2_score(y_test, predicciones_lineal)

print("Regresión Lineal")
print("MAE:", mae_lineal)
print("RMSE:", rmse_lineal)
print("R²:", r2_lineal)

Regresión Lineal
MAE: 970043.4039201637
RMSE: 1324506.9600914384
R²: 0.6529242642153185


In [10]:
modelo_arbol = DecisionTreeRegressor(random_state=42)

# Creo el árbol de decisión porque quiero comparar otro modelo
# que pueda tomar decisiones a partir de las características de las casas.

modelo_arbol.fit(X_train, y_train)

# Entreno el árbol utilizando los mismos datos que utilicé
# para entrenar la regresión lineal.

predicciones_arbol = modelo_arbol.predict(X_test)

# Utilizo el árbol entrenado para predecir los precios
# de las casas que están en los datos de prueba.

In [11]:
mae_arbol = mean_absolute_error(y_test, predicciones_arbol)
rmse_arbol = np.sqrt(mean_squared_error(y_test, predicciones_arbol))
r2_arbol = r2_score(y_test, predicciones_arbol)

print("Árbol de Decisión")
print("MAE:", mae_arbol)
print("RMSE:", rmse_arbol)
print("R²:", r2_arbol)

Árbol de Decisión
MAE: 1195266.0550458715
RMSE: 1625669.904259373
R²: 0.4771459275854347


In [12]:
resultados = pd.DataFrame({
    'Modelo': ['Regresión Lineal', 'Árbol de Decisión'],
    'MAE': [mae_lineal, mae_arbol],
    'RMSE': [rmse_lineal, rmse_arbol],
    'R²': [r2_lineal, r2_arbol]
})

# Creo una tabla para comparar los resultados de los dos modelos
# y así identificar cuál tuvo un mejor desempeño.

resultados

,Modelo,MAE,RMSE,R²
0,Regresión Lineal,9.700434e+05,1.324507e+06,0.652924
1,Árbol de Decisión,1.195266e+06,1.625670e+06,0.477146


## Análisis y conclusión

Al comparar los dos modelos, la Regresión Lineal presentó mejores resultados que el Árbol de Decisión.

La Regresión Lineal obtuvo un MAE de 970043, un RMSE de 1324507 y un R² de 0,6529. Por otro lado, el Árbol de Decisión obtuvo un MAE de 1195266, un RMSE de 1625670 y un R² de 0,4771.

Considero que la Regresión Lineal fue más adecuada para este conjunto de datos porque presentó menores errores y un R² mayor. Esto indica que logró representar mejor la relación entre las características de las casas y sus precios.

## Conclusión

En este ejercicio desarrollé dos modelos predictivos para estimar el precio de las casas: Regresión Lineal y Árbol de Decisión. Después de entrenarlos y evaluarlos con los mismos datos de prueba, la Regresión Lineal obtuvo el mejor desempeño.

Para este conjunto de datos, considero que la Regresión Lineal es el modelo más adecuado, aunque sus predicciones todavía presentan cierto margen de error.